pkg/planner/core/hint_utils.go

```diff
@@ -220,59 +312,185 @@ func getTableName(tblName model.CIStr, asName *model.CIStr) model.CIStr {
+// It generates a join method hint for a single physical join operator according to the input joinType.
+// Both children of the Join should be passed in as the children arguments, this is for correctly deriving the QB offset
+// for the hint.
+// For hints like merge_join(), we can generate hint using table name of any one of the two tables. But for hints like
+// hash_join_build() and inl_join(), we want to generate hint using table name of a specific side. For this difference,
+// we introduce the onlyFirstTbl argument. If onlyFirstTbl is true, we only try to generate hint using the table name of
+// the children[0].
+func genJoinMethodHintForSinglePhysicalJoin(
+	sctx base.PlanContext,
+	joinType string,
+	parentOffset int,
+	nodeType h.NodeType,
+	onlyFirstTbl bool,
+	children ...base.PhysicalPlan,
+) *ast.TableOptimizerHint {
 	if parentOffset == -1 {
-		return res
+		return nil
+	}
+	hintTbls, hintQBName := genHintTblForJoinNodes(sctx, children, parentOffset, nodeType)
+	effectiveHintTbls := slices.DeleteFunc(slices.Clone(hintTbls), func(ht *ast.HintTable) bool {
+		return ht == nil
+	})
+	if len(effectiveHintTbls) == 0 {
+		return nil
+	}
+
+	if onlyFirstTbl && hintTbls[0] == nil {
+		return nil
+	}
+
+	newHint := &ast.TableOptimizerHint{
+		HintName: model.NewCIStr(joinType),
+		Tables:   []ast.HintTable{*effectiveHintTbls[0]},
 	}
-	for _, child := range children {
-		qbOffset, ht := extractHintTableForJoinNode(sctx, child, parentOffset)
+
+	if hintQBName != nil {
+		newHint.QBName = *hintQBName
+	}
+
+	return newHint
+}
+
+// genHintTblForJoinNodes tries to generate ast.HintTable for each join node, and the QB name for the hint itself.
+// (Join node here means the operators that are joined, not Join operator itself)
+// If the return values is not (nil,nil), len(hintTbls) should be equal to len(joinedNodes). The invalid ones in the
+// returned hintTbls slice will be nil.
+// The hintQBNamePtr will be nil if it's not needed, or we failed to generate one.
+func genHintTblForJoinNodes(
+	sctx base.PlanContext,
+	joinedNodes []base.PhysicalPlan,
+	parentOffset int,
+	nodeType h.NodeType,
+) (hintTbls []*ast.HintTable, hintQBNamePtr *model.CIStr) {
+	// 1. Use genHintTblForSingleJoinNode() to generate QB offset and table name for each join node.
+
+	// Note that if we failed to generate valid information for one element in joinedNodes, we append -1 and nil instead
+	// of skipping.
+	// So qbOffsets[x] is -1 if and only if hintTbls[x] is nil;
+	// and qbOffsets[x] >=0 if and only if hintTbls[x] is not nil.
+	hintTbls = make([]*ast.HintTable, 0, len(joinedNodes))
+	qbOffsets := make([]int, 0, len(joinedNodes))
+	guessQBOffsets := make(map[int]struct{})
+	for _, plan := range joinedNodes {
+		qbOffset, guessOffset, ht := genHintTblForSingleJoinNode(sctx, plan, parentOffset)
 		if qbOffset < 0 || ht == nil {
+			qbOffsets = append(qbOffsets, -1)
+			hintTbls = append(hintTbls, nil)
```

Comment:
Should we "continue" after these 2 append operations?

Reply:
Good catch. Updated.